### PR TITLE
112899357_refactor_case_statement_in_authentication_steps

### PR DESCRIPTION
**Pull request by decareano**
First attempt to refactor "Case" statement in `features/step_definitions/authentication_steps.rb`. 
Following Sam's suggestions to use Procs/Blocks. Issue: the procs are different. So in order to get the ball rolling with the refactoring, I only added Poltergeist (headless) and Selenium (non-headless). I would like some ideas as to how to make the "Given" work with WebKit and RackTest drivers.

### CODE REVIEW

**Comment by tansaku** [comment]
hi @decareano thanks for this.  I pulled down and got it sorted like so:

```rb
PoltergeistProc = -> (page, key, value) do
  page.driver.set_cookie(key, value)
end

SeleniumProc = -> (page, key, value) do
  page.driver.browser.manage.add_cookie(name: key, value: value)
end

RackProc = -> (_page, key, value) do
  headers = {}
  Rack::Utils.set_cookie_header!(headers, key, value)
  cookie_string = headers['Set-Cookie']
  Capybara.current_session.driver.browser.set_cookie(cookie_string)
end

WebkitProc = -> (_page, key, value) do
  headers = {}
  Rack::Utils.set_cookie_header!(headers, key, value)
  cookie_string = headers['Set-Cookie']
  Capybara.current_session.driver.browser.set_cookie(cookie_string)
end

hash = {
    Capybara::Selenium::Driver => SeleniumProc,
    Capybara::Poltergeist::Driver => PoltergeistProc,
    Capybara::RackTest::Driver => RackProc,
    Capybara::Webkit::Driver => WebkitProc
}

Given /^I have a "([^\"]+)" cookie set to "([^\"]+)"$/ do |key, value|
  proc = hash[Capybara.current_session.driver.class]
  raise "no cookie for driver #{Capybara.current_session.driver.class.name}" if proc.nil?
  proc.call(page, key, value)
end
```

I'm not saying this is the best way to do, but this passes the tests, and just gets it working in the way I was trying to suggest.  Unsurprisingly there was a minor bug in the code I put in the PT ticket - specifically that we need to look up in the hash via the class entity.

I was going to push this code direct to your repo, but I don't have permission - please feel free to add me as a contributor as it allows me to fix up your pull requests directly - but failing that I can re-open them like I've done here:

https://github.com/AgileVentures/LocalSupport/pull/349

So the things we would want to look at going forward would be:

* [ ] better place for this code (as you suggested)
* [ ] should we use this code at all?  perhaps we should prefer https://github.com/nruth/show_me_the_cookies
* [ ] which of these drivers do we actually use regularly
* [ ] should we run a complete test suite in selenium to see if that's still a relevant debug option 
* [ ] refactor further? should the procs actually be classes? probably not since they don't manage any state

if https://github.com/nruth/show_me_the_cookies is solid and stable then that's probably a better option than any code we can hand roll ...

**Comment by tansaku** [comment]
@decareano so I think your develop branch is maybe out of date - you might want to delete that and re-pull it from upstream; then cherry-pick these latest changes on top of that.  I can do that for you, or would you like to pair on this at some point?

**Comment by decareano** [comment]
@sam, if you merged code in the last week, my *develop* is out of date. I will re-pull from upstream. But last time I tried to cherry-pick, it did not work. can you provide me with a quick how-to in a few steps? Also, what's the meaning of `= ->` in the procs?

**Comment by tansaku** [comment]
hi @decareano (note that I'm @tansaku on github and not @ sam :-) )

assuming that your current develop is relatively up to date (I was just looking and seeing only new commits in this PR) then you should be able to get away without cherry-picking, perhaps just doing:

```sh
git fetch upstream
git checkout develop
git pull upstream/develop
```

to get your develop up to date and then update the PR branch like so

```sh
git checkout 112899357_refactor_case_statement_in_authentication_steps
git rebase develop
git push origin 112899357_refactor_case_statement_in_authentication_steps
```

and it might be okay like that ...

and here's info on the `->` short lambda syntax: https://makandracards.com/makandra/17305-short-lambda-syntax-in-ruby-1-9

**Comment by decareano** [comment]
1. Sam wrote: better place for this code (as you suggested) :
_I think you put the procs in the correct place already (`features\support\helpers.rb`)_ should they go somewhere else? 
2. Sam wrote: should we use this code at all? perhaps we should prefer https://github.com/nruth/show_me_the_cookies:
_I am going to work on the other items first and then circle back to the gem if you don't mind._
3. Sam wrote: which of these drivers do we actually use regularly
should we run a complete test suite in selenium to see if that's still a relevant debug option
refactor further? :
_I want to understand this better. My understanding of how JS works with Rails is not that great. I am looking at the RailsGuides for JS and Rails right now. I welcome suggestions for reading. And then there is the browser, the drivers and cucumber tests. Lots of moving parts. By running Selenium you mean take out all the tags and let it go at it ?_ 
4. should the procs actually be classes? probably not since they don't manage any state : 
_I am under the impression that procs are objects that has been bound to a set of local variables and thus can be called in different contexts and be able to access those local variables._

File: features/support/helpers.rb [hunk 1]
```diff
@@ -44,6 +44,36 @@ def stub_request_with_address(address, body = nil)
...
+  cookie_string = headers['Set-Cookie']
+  Capybara.current_session.driver.browser.set_cookie(cookie_string)
+end
+
+WebkitProc = -> (_page, key, value) do
+  headers = {}
+  Rack::Utils.set_cookie_header!(headers, key, value)
+  cookie_string = headers['Set-Cookie']
+  Capybara.current_session.driver.browser.set_cookie(cookie_string)
+end
```
Comment (johnnymo87) [inline features/support/helpers.rb]: Does everything still work if you make `WebkitProc` just an alias of `RackProc`, i.e. `WebkitProc = RackProc`?

Comment (tansaku) [inline features/support/helpers.rb]: probably - great suggestion @johnnymo87 !

**Comment by tansaku** [comment]
@decareano wrote

> 1. Sam wrote: better place for this code (as you suggested) :
> _I think you put the procs in the correct place already (`features\support\helpers.rb`)_ should they go somewhere else? 

Yeah, helpers is probably fine for now - although that helpers file is getting full of stuff - they could go in their own file (module?) but that's low priority compared to assessing whether we can use the gem.

> 2. Sam wrote: should we use this code at all? perhaps we should prefer https://github.com/nruth/show_me_the_cookies:
> _I am going to work on the other items first and then circle back to the gem if you don't mind._

Well ultimately it's up to you, but if it was me I would focus on the gem first, since if it works as we want then it would make the other work here redundant.  I guess it depends on what our objective is.  I'd love to avoid working on something that we're then going to throw out, but sometimes that's useful for learning, and perhaps learning is your priority.  We could look at the gem and then still work on our custom version for learning purposes.

Anyway, all I can say is that if I was taking the lead on this activity I'd want to get as much information about the gem operation and the state of the community maintaining it before doing anything else.  But that's me - you've got to do what you're comfortable with.  As you say below there are a lot of moving parts here.

Maybe I should take the job of looking at the gem and tell you what I find out ...?

> 3. Sam wrote: which of these drivers do we actually use regularly
should we run a complete test suite in selenium to see if that's still a relevant debug option
refactor further? :
> _I want to understand this better. My understanding of how JS works with Rails is not that great. I am looking at the RailsGuides for JS and Rails right now. I welcome suggestions for reading. And then there is the browser, the drivers and cucumber tests. Lots of moving parts. By running Selenium you mean take out all the tags and let it go at it ?_ 

No, I mean that we'd change (temporarily) line 35 in `features/support/en.rb` from:

```rb
Capybara.javascript_driver = :webkit
```
to 
```rb
Capybara.javascript_driver = :selenium
```

And this would allow us to run alot of the cucumber tests in a real popup browser.  As I mentioned we need to do that occasionally for test/debugging purposes.

I empathise with you confronting the complexity here - there are a lot of moving parts as you say.  Here are some things you might like to read on how capybara javascript testing works:

* https://robots.thoughtbot.com/write-reliable-asynchronous-integration-tests-with-capybara
* http://tutorials.jumpstartlab.com/topics/capybara/capybara_with_selenium_and_webkit.html

To understand more about how Rails works with JavaScript you might want to read:

* http://guides.rubyonrails.org/asset_pipeline.html
* http://stackoverflow.com/a/15872400/316729

You might also want to try and set up a pairing session with me - I think pairing with people on these things and getting them working together is the most powerful and effective way to learn.

> 4. should the procs actually be classes? probably not since they don't manage any state : 
> _I am under the impression that procs are objects that has been bound to a set of local variables and thus can be called in different contexts and be able to access those local variables._

The procs are objects.  As you know, everything in Ruby is an object.  We are assigning each proc to a global variable (e.g. SeleniumProc).  Actually CamelCase is reserved in Ruby for classes and modules, so we should probably use ALL_CAPS, e.g. SELENIUM_PROC to highlight that.  Procs get access to the local (and global) variables that are accessible when they are created.  The selenium proc as it currently stands only accesses the arguments that it is passed (i.e. page, key and value).  RackProc and WebkitProc refer to variables Rack and Capybara.

As you say the proc can be called in different contexts.  We can pass in different arguments to modify what the proc does, but local variables in the new context should not have any impact on the operation of the proc (I think).  A proc is a "closure" in that it closes over the values of any local variables.  Changes to those variables in the new context should not impact the proc (I think).  We're not using procs here because of closures though, or anything to do with contexts.  We're just using them because they are simpler than classes, and we don't need the complexity of a class.

Procs are just stand-alone methods that we can assign to a variable and then pass around.  Does that make sense?  Perhaps I should write a class based version so you can see an alternative ...?

**Comment by tansaku** [comment]
@decareano I took the liberty of syncing this with develop for you :-)

**Comment by decareano** [comment]
+1

**Comment by decareano** [comment]
@tansaku, so you sync my branch to develop. Fine. But when I fetch from develop and try to checkout my branch (112899357_refactor_case_statement_in_authentication_steps) it does not work. it returns the typical error: <branch name> did not match any files know to git.

File: features/support/helpers.rb (hunk 1 above)
Comment (Maroo-b) [inline features/support/helpers.rb]: Nice idea @johnnymo87, but I saw that we are using a Hash to get the correct proc, why not set the value of RackTest driver and WebKit driver to the same proc for example RackProc ?

**Comment by tansaku** [comment]
@decareano that may be because of the history tampering we did with develop after those failing tests were checked in.  

I think you might want to delete your local versions of develop and this branch, and then pull them back in.

```sh
$ git branch -D develop
$ git branch -D 112899357_refactor_case_statement_in_authentication_steps 
$ git fetch upstream
$ git fetch origin
$ git checkout develop
$ git checkout 112899357_refactor_case_statement_in_authentication_steps
```

assuming that you've appropriated recorded or stashed any outstanding AND important local changes you have

**Comment by tansaku** [comment]
closing as we're going with the gem instead

**Pull request closed without merging**